### PR TITLE
Issue/7154 compacted env vars

### LINKED ISSUE #7154: Allow compacted env_vars
1. add a field to the compile request for compactable environment variables. This will be a dict similar to the non-mergeable environment variables.
2. When merging of compiles, everything remains as before, except that we merge the compactable env variables toghether  (space separated list) 
3. this may require some change to what we output, to keep users aware of the env vars passed in
4. Make sure we never merge beyond the limit of a single environment variable.

Come up with better name. Best proposal so far: mergeable.

e.g.
```python
CompileRequest(
  env_vars={
    "environment_id": "abc",
  },
  mergeable_env_vars={
    "instance_id": 123,
  },
  ...
)
```

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

Add the ability to have mergeable environment variables 
closes #7154 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
```
src/inmanta/data/__init__.py:5048: error: Argument 1 to "loads" has incompatible type "object"; expected "str | bytes | bytearray"  [arg-type]
src/inmanta/data/dataview.py:1122: error: Argument 1 to "loads" has incompatible type "object"; expected "str | bytes | bytearray"  [arg-type]
src/inmanta/data/dataview.py:1254: error: Argument 1 to "loads" has incompatible type "object"; expected "str | bytes | bytearray"  [arg-type]

```
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )
- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)

### CODE REVIEW

File: tests/server/test_compilerservice.py [hunk 1]
```diff
@@ -993,27 +995,34 @@ async def test_compileservice_queue(mocked_compiler_service_block: queue.Queue,
     remote_id6 = uuid.uuid4()
     compile_id6, _ = await compilerslice.request_recompile(env=env, force_update=False, do_export=True, remote_id=remote_id6)
 
+    # request with partial, will no be merged
```
Comment (arnaudsjs) [inline tests/server/test_compilerservice.py:998]:
```suggestion
    # request with partial, will not be merged
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/server/services/compilerservice.py [hunk 2]
```diff
@@ -832,9 +842,20 @@ async def _run(self, compile: data.Compile) -> None:
 
         runner = self._get_compile_runner(compile, project_dir=os.path.join(self._env_folder, str(compile.environment)))
 
+        merged_env_vars = dict(compile.mergeable_environment_variables)
+        for merge_candidate in merge_candidates:
+            for key, to_add in merge_candidate.mergeable_environment_variables.items():
+                existing = merged_env_vars.get(key, None)
+                if existing:
+                    merged_env_vars[key] = existing + " " + to_add
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/compilerservice.py:850]: It is documented as a space separated list, so if you have spaces in the value, it is always a problem.
If you want them escaped, you should escape them I think.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py [hunk 3]
```diff
@@ -3958,6 +3985,14 @@ def to_dto(self) -> m.CompileRun:
             failed_compile_message=self.failed_compile_message,
         )
 
+    def to_dict(self) -> JsonType:
+        """produce dict directly, for untyped endpoints"""
+        # mangle the output for backward compatibility
+        # we have to do it because we have no DTO here
+        out = super().to_dict()
+        out["environment_variables"] = self.used_environment_variables or self.requested_environment_variables
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/__init__.py:3993]: dang, good point

File: src/inmanta/data/dataview.py [hunk 4]
```diff
@@ -683,7 +684,15 @@ def construct_dtos(self, records: Sequence[Record]) -> Sequence[model.CompileRep
                 do_export=compile["do_export"],
                 force_update=compile["force_update"],
                 metadata=json.loads(compile["metadata"]) if compile["metadata"] else {},
-                environment_variables=json.loads(compile["environment_variables"]) if compile["environment_variables"] else {},
+                environment_variables=(
+                    json.loads(compile["used_environment_variables"]) if compile["used_environment_variables"] else {}
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/data/dataview.py:688]: moved to validator

**Review by FloLey — APPROVED** [review]

File: src/inmanta/data/dataview.py [hunk 5]
```diff
@@ -683,7 +684,11 @@ def construct_dtos(self, records: Sequence[Record]) -> Sequence[model.CompileRep
...
```
Comment (arnaudsjs) [inline src/inmanta/data/dataview.py:688]:
```suggestion
                    json.loads(compile["used_environment_variables"]) if compile["used_environment_variables"] else None
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by arnaudsjs — APPROVED** [review]

**Pull request closed without merging**